torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1979,6 +1983,906 @@ def clip_grad_norm_(
+            for param in fsdp_module.params:  # may be empty
+                flat_param_to_fsdp_module[param] = fsdp_module
+
+        group = model.process_group if hasattr(model, "process_group") \
+            else None  # not all `torch.nn.Module`s have `process_group`
+        rank = dist.get_rank(group)
+        to_save = rank == target_rank
+        full_osd: Dict = {"state": {}, "param_groups": []} if to_save else {}
+        full_osd_state = full_osd["state"] if to_save else None  # alias
+
+        # Save a mapping from flattened to unflattened parameter IDs to remap
+        # the IDs in the "param_groups" part
+        flat_to_unflat_param_ids: Dict[int, List[int]] = {}
+        unflat_param_id = 0
+
+        # Handle the "state" part of the optimizer state dict
+        flat_param_id_to_param = \
+            FSDP._get_flat_param_id_to_param(model, optim_input)
+        for flat_param_id, param in enumerate(flat_param_id_to_param):  # type: ignore[assignment]
+            if flat_param_id not in flat_to_unflat_param_ids:
+                flat_to_unflat_param_ids[flat_param_id] = []
+            unflat_param_ids = flat_to_unflat_param_ids[flat_param_id]  # alias
+            # Do not include parameters without state to avoid empty mappings
+            if flat_param_id not in osd_state:
+                unflat_param_ids.append(unflat_param_id)
+                unflat_param_id += 1
+                continue
+            if isinstance(param, FlatParameter):
+                assert param in flat_param_to_fsdp_module, \
+                    "Check the `flat_param_to_fsdp_module` mapping " \
+                    f"construction\nparam={param}"
+                fsdp_module = flat_param_to_fsdp_module[param]
+                flat_param_state = osd_state[flat_param_id]
+                unflat_state = FSDP._unflatten_optim_state(
+                    param, flat_param_state, fsdp_module, target_rank,
+                )
+                for unflat_param_state in unflat_state.values():
+                    if to_save:
+                        full_osd_state[unflat_param_id] = unflat_param_state
+                    unflat_param_ids.append(unflat_param_id)
+                    unflat_param_id += 1
+            else:  # do not need to unflatten
+                if to_save:
+                    # Do not `deepcopy()` to avoid unnecessarily duplicating
+                    # tensor storage
+                    full_osd_state[unflat_param_id] = \
+                        copy.copy(osd_state[flat_param_id])
+                    # Move all tensor state to CPU
+                    for state_name, value \
+                            in full_osd_state[unflat_param_id].items():
+                        if torch.is_tensor(value):
+                            full_osd_state[unflat_param_id][state_name] = \
+                                value.cpu()
+                unflat_param_ids.append(unflat_param_id)
+                unflat_param_id += 1
```

Comment:
Because we are calling this method on a root module that may not be an `FSDP` instance, we must account for other parameters that are not `FlatParameter`s. This code path is to handle those.

Reply:
I see, thanks for explaining, would you please add a comment for this as well?